### DISCUSSION TITLE
RDKit for Pyodide

### DISCUSSION THREAD

**Discussion by skalkin** [rdkit/rdkit#7827] [Ideas]
Pyodide is a Python distribution for the browser and Node.js based on WebAssembly. At [Datagrok](https://datagrok.ai/), we have recently integrated with Pyodide, and you can now execute Python scripts in the browser (server-side scripting for Python/R/Julia/etc [was available for a long time](https://datagrok.ai/help/compute/scripting/), also [some RDKit Python scripts](https://datagrok.ai/help/datagrok/solutions/domains/chem/scripts/gasteiger-charges)). This is very convenient for data scientists who do basic data wrangling (like post-processing of SQL queries) and other lightweight scripting, as they prefer Python over JS for obvious reasons. A number of popular packages such as pandas, scipy, and scikit-learn are [available out of the box](https://pyodide.org/en/stable/usage/packages-in-pyodide.html).

We have a sizable community of cheminformaticians that use Datagrok, and of course they would love to use RDKit from Python, but RDKit is not yet available as a Pyodide package. Pure Python packages work out of the box, and you can [create a C++ based Pyodide package](https://pyodide.org/en/stable/development/new-packages.html) as well. Hopefully this won't be too hard, since we already have WASM-based MinimalLib.

We might decide to build the Pyodide version of RDKit, but before we start it I hope to get some feedback from the community. Would it be useful to others? Any issues with the build/sync we are likely to encounter? Any other feedback is very welcome, too :)

**Comment by ShangChien** [comment]
👋 Nice see the needs and direction of your team. This has a lot of potential.

**Comment by seandearnaley** [comment]
this would be really great if it were able to expose the maximal lib in the browser or node environment

**Comment by skalkin** [comment]
This would really be a great addition, I can think of hundreds of good use cases for it. Unfortunately I might not have the right technical skills to tackle this, but let's see if we can help make it happen by connecting people :)

I spoke to Gyeongjae Choi @ryanking13, one of the maintainers of Pyodide, and he is willing to help (thanks a lot Gyeongjae!), but he needs some support from RDKit's maintainers. @greglandrum , @ptosco - I wonder if there is something you can help with? Please see below our chat with Gyeongjae:


> Regarding RDkit, I'm not working on builds for individual packages as I've been prioritizing other work lately, but if anyone has knowledge of rdkit and is interested in building it, I can help with build issues.

> Andrew:
>  Thanks a lot! I’ll try to find the right person to do it. Any idea how big of an effort it might be? 

> It depends on how complex the RDkit's build system is. One issue I found when reading the RDKit documentation is that it does not use a standard Python build system (PEP 517), instead it directly calls CMake commands to build the package. The Pyodide build system doesn't fit this kind of package very well.

> A similar case is the RobotRaconteur package, which also doesn't follow the standard Python build system. However, for RobotRaconteur, the package maintainer is directly involved in building the package and has resolved the issues. I think RDKit is a similar case where someone who understands the package's build system should be able to resolve it.

**Comment by ptosco** [comment]
@skalkin Hi Andrew, here's a few considerations:
* The standard RDKit Python wrappers depend on the `boost::python` library, which I doubt can be easily built with `emscripten`. Plus, the full WASM RDKit library would be _very_ large
* The MinimalLib itself currently does not have Python wrappers. Therefore, to expose MinimalLib to Python one would need to write dedicated Python wrappers, and build them with `emscripten` just for Pyodide. This would require significant effort. Additionally, these wrappers would only cover the limited functionality exposed through MinimalLib. Any RDKit user who is familiar with the standard RDKit Python wrappers would rather expect to find the full RDKit functionality, and would probably be disappointed finding only a subset of the library with a very different interface, which would need to be learnt from scratch

A simpler alternative might be to call the existing JS MinimalLib interface from Python (https://pyodide.org/en/stable/usage/type-conversions.html#calling-javascript-functions-from-python).
This would at least not require building a new WASM library just for Pyodide, but would still have the drawback that it would only cover MinimalLib. And I can imagine would come with the usual challenges of C++ (rather, WASM) object lifetime management within Python.

**Comment by ryanking13** [comment]
Hi @ptosco. Thanks for the reply. Here are my 2 cents as a Pyodide maintainer.

> The standard RDKit Python wrappers depend on the boost::python library, which I doubt can be easily built with emscripten. 

We (Pyodide) [do build boost to WASM](https://github.com/pyodide/pyodide/blob/34fcd02172895d75db369994011409324f9e3cce/packages/boost-cpp/meta.yaml), but didn't enable boost.Python. I think it won't be very hard to enable boost.Python but I'll need to check.

> Plus, the full WASM RDKit library would be very large

How large is it (full distribution, minimum distribution)? The RDKit wheel in the PyPI is around 30MB. We already have packages like SciPy, DuckDB, Polars, PyArrow that are larger than 30MB, so probably the size isn't that concern.

> The MinimalLib itself currently does not have Python wrappers. Therefore, to expose MinimalLib to Python one would need to write dedicated Python wrappers, and build them with emscripten just for Pyodide.

I wonder how MinimalLib (or rdkit-js) is built. Is it a thin wrapper of RDKit library? or does it contain several other features as well? If the former is the case, I guess the right direction is not to expose MinimalLib, but instead build the RDKit directly to Pyodide (using the Python build system)

> A simpler alternative might be to call the existing JS MinimalLib interface from Python (https://pyodide.org/en/stable/usage/type-conversions.html#calling-javascript-functions-from-python).

Yes, this is also an option. Since Pyodide can call JS functions, one should be able to use RDKit in Pyodide today by calling MinimalLib interfaces, but the downside is this that it will require some code modifications for people who are using the existing Python RDKit library, which will cause some inconvenience.

**Reply by greglandrum** [reply]
@ryanking13 I'm sure that you probably know this already, but build systems for complex pieces of software are themselves complex (and, often, fragile). Just getting the python build system to build the RDKit wrappers using the standard C++ toolchain (never mind using empscripten) is probably a deeply non-trivial task.

I think it would be cool if there were a robust and supportable way to make the RDKit available via Pyodide but, at least from where I sit, it seems likely to be a serious amount of work.

**Reply by ryanking13** [reply]
Thanks for the reply. Yes, I understand how complex the build system is for large software. I agree that, if it would require large changes or custom patches to build RDKit to Pyodide, it wouldn't be a sustainable way.

Our goal with Pyodide (and Emscripten) is to leverage the existing build system (which targets the Linux platform) while minimizing the extra processing. Of course, as libraries get larger, this is often not possible and requires a lot of package maintainer involvement. So, our approach is not to push these requests to package maintainers but rather to support them if they are interested in building packages for Pyodide.

**Comment by stefdoerr** [comment]
Since the discussion has been stale for a year now I attempted it myself. I managed to build rdkit for pyodide: https://github.com/rdkit/rdkit/pull/9181
Now I decided to try to add it to the pyodide-recipes repo instead where it seems more fitting. If I have success I'll let you know.

**Reply by flange-ipb** [reply]
> Now I decided to try to add it to the pyodide-recipes repo instead where it seems more fitting.

There happens to be quite some progress by @fxcoudert in https://github.com/pyodide/pyodide-recipes/pull/509.

**Reply by stefdoerr** [reply]
Oh! Somehow my googling missed it. Seems like it's not as feature-complete yet as the one which I built in the PR. I will give it a quick try anyway, and if I fail I'll leave it to @fxcoudert

**Reply by fxcoudert** [reply]
I have not managed a full build so far, so if you have managed to build it, it's great! I keep getting errors about parts being built as shared objects instead of static.

**Reply by stefdoerr** [reply]
I think I got it working just now based on my previous PR :) I'll make a pyodide-recipes PR tomorrow.

**Reply by fxcoudert** [reply]
Amazing! That will make teaching so much easier.

**Reply by fxcoudert** [reply]
Could the rdkit cmake support be amended to make the wasm compilation more transparent? It seems like you have to hack around several issues in your build. Is there a way to also improve the upstream support for these builds?

**Reply by stefdoerr** [reply]
Done now! The previous PR had a few more hacks than were strictly necessary. You can take a look here at my PR to pyodide-recipes https://github.com/pyodide/pyodide-recipes/pull/538 
But I would agree that upstream could probably take some inspiration from those hacks to improve the cmake building, I would have given up if I was writing this manually.

**Comment by stefdoerr** [comment]
https://github.com/pyodide/pyodide-recipes/pull/538#issuecomment-4087724101
Any opinions from the rdkit devs? Would you be able to host the CI for the pyodide wheels?
